drivers/adc/adc_handlers.c

```diff
@@ -23,21 +24,46 @@ Z_SYSCALL_HANDLER(adc_disable, dev)
 
 Z_SYSCALL_HANDLER(adc_read, dev, seq_table_p)
 {
-	struct adc_seq_entry *entry;
+	struct adc_seq_entry *entry, *entries_copy;
 	struct adc_seq_table *seq_table = (struct adc_seq_table *)seq_table_p;
-	int i;
+	struct adc_seq_table seq_table_copy;
+	unsigned int entries_bounds;
+	int i = 0, ret;
 
 	Z_OOPS(Z_SYSCALL_DRIVER_ADC(dev, read));
 	Z_OOPS(Z_SYSCALL_MEMORY_READ(seq_table, sizeof(struct adc_seq_table)));
-	Z_OOPS(Z_SYSCALL_MEMORY_ARRAY_READ(seq_table->entries,
-					   seq_table->num_entries,
-					   sizeof(struct adc_seq_entry)));
 
-	for (entry = seq_table->entries, i = 0; i < seq_table->num_entries;
+	seq_table_copy = *seq_table;
+	if (Z_SYSCALL_VERIFY_MSG(
+			!__builtin_umul_overflow(seq_table_copy.num_entries,
+						 sizeof(struct adc_seq_entry),
+						 &entries_bounds),
+			"num_entries too large")) {
+		ret = -EINVAL;
```

Comment:
Are we returning signed integers from syscalls now?

Reply:
The generated function for the system call invocation will be signed.
Stuff gets cast to unsigned internally but it should still pop out as expected.
Otherwise, many other syscalls that return negative values would not work.